### DISCUSSION TITLE
Hitting abuse rate limit in github action when deleting tags

### DISCUSSION THREAD

**Discussion by jamie-wearsafe** [octokit/octokit.js#1980] [Q&A] [answered]
Hi,
In my GitHub action, I am wanting to delete a large number of tags and releases that match a regex, I have found that when there are a large number of results I get the error:

```sh
Delete release error: HttpError: You have triggered an abuse detection mechanism. Please wait a few minutes before you try again.
```

The code is here:
https://github.com/Broadshield/github-action-cleanup-releases-and-tags

Is there a way to run this action without hitting the limits?

**Comment by gr2m** [comment, marked as answer]
It doesn't look like you are using `@octokit/rest`, but `actions-toolkit`. I do not maintain this library. I think they use Octokit internally but I don't know how up-to-date it is. If you like you can use https://github.com/octokit/action.js, in which case you could load the [throttle plugin](https://github.com/octokit/plugin-throttling.js/), which will help to reduce the problem of hitting rate/abuse limits.

```js
const { Octokit } = require("@octokit/action");
const { throttling } = require("@octokit/plugin-throttling");

const MyOctokit = Octokit.plugin(throttling).defaults({
    throttle: {
    onRateLimit: (retryAfter, options, octokit) => {
      octokit.log.warn(
        `Request quota exhausted for request ${options.method} ${options.url}`
      );

      if (options.request.retryCount === 0) {
        // only retries once
        octokit.log.info(`Retrying after ${retryAfter} seconds!`);
        return true;
      }
    },
    onAbuseLimit: (retryAfter, options, octokit) => {
      // does not retry, only logs a warning
      octokit.log.warn(
        `Abuse detected for request ${options.method} ${options.url}`
      );
    },
  },
})
```

Then instead of using `tools.github`, use `octokit`

```js
const octokit = new MyOctokit()
```

**Reply by jamie-wearsafe** [reply]
I updated the code with the suggestions you made: https://github.com/Broadshield/github-action-cleanup-releases-and-tags

Above you mention `require("@octokit/plugin-throttle");` but the plugin readme says to use `require("@octokit/plugin-throttling");` incase anyone else reads this.

```js
const { throttling } = require("@octokit/plugin-throttling");

const MyOctokit = Octokit.plugin(throttling).defaults({
```

**Reply by gr2m** [reply]
Glad that it helped!

> Above you mention `require("@octokit/plugin-throttle");` but the plugin readme says to use `require("@octokit/plugin-throttling");` incase anyone else reads this.

I fixed it in my answer, thanks!

**Reply by jamie-wearsafe** [reply]
Additionally, I was looking for a way to get the Action's context, inputs and set outputs, and context without using @actions/github, is there a builtin way to do this in octokit?

**Reply by gr2m** [reply]
For that I would recommend to use [`@actions/core`](https://github.com/actions/toolkit/tree/main/packages/core#readme)

You can get the event payload with

```js
const eventPayload = require(process.env.GITHUB_EVENT_PATH);
```

Here is the list of all available environment variables:
https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables
